okapi-core/src/test/java/org/folio/okapi/util/DepResolutionTest.java

```diff
@@ -220,23 +235,180 @@ public void test4(TestContext context) {
+      context.assertTrue(res.succeeded());
+      logger.debug("tml result = " + Json.encodePrettily(tml));
+      context.assertEquals(1, tml.size());
+      context.assertEquals("moduleD-1.0.0", tml.get(0).getId());
+      context.assertEquals(null, tml.get(0).getFrom());
+      context.assertEquals("enable", tml.get(0).getAction().name());
+      async.complete();
+    });
+  }
+
+  // install optional with existing interface that is too low (error)
+  @Test
+  public void testInstallOptionalFail(TestContext context) {
+    Async async = context.async();
+
+    Map<String, ModuleDescriptor> modsAvailable = new HashMap<>();
+    modsAvailable.put(mdA100.getId(), mdA100);
+    modsAvailable.put(mdD100.getId(), mdD100);
+    modsAvailable.put(mdD110.getId(), mdD110);
+    modsAvailable.put(mdE100.getId(), mdE100);
+
+    Map<String, ModuleDescriptor> modsEnabled = new HashMap<>();
+    modsEnabled.put(mdA100.getId(), mdA100);
+
+    List<TenantModuleDescriptor> tml = new LinkedList<>();
+    TenantModuleDescriptor tm = new TenantModuleDescriptor();
+    tm.setAction(TenantModuleDescriptor.Action.enable);
+    tm.setId(mdD110.getId());
+    tml.add(tm);
+
+    DepResolution.installSimulate(modsAvailable, modsEnabled, tml, res -> {
+      context.assertTrue(res.failed());
+      context.assertEquals("enable moduleD-1.1.0 failed: interface int required by module moduleD-1.1.0 not found", res.cause().getMessage());
+      async.complete();
+    });
+  }
+
+  // install optional with existing interface that needs upgrading
+  @Test
+  public void testInstallOptionalExistingModule(TestContext context) {
+    Async async = context.async();
+
+    Map<String, ModuleDescriptor> modsAvailable = new HashMap<>();
+    modsAvailable.put(mdA100.getId(), mdA100);
+    modsAvailable.put(mdA110.getId(), mdA110);
+    modsAvailable.put(mdD100.getId(), mdD100);
+    modsAvailable.put(mdD110.getId(), mdD110);
+    modsAvailable.put(mdE100.getId(), mdE100);
+
+    Map<String, ModuleDescriptor> modsEnabled = new HashMap<>();
+    modsEnabled.put(mdA100.getId(), mdA100);
+
+    List<TenantModuleDescriptor> tml = new LinkedList<>();
+    TenantModuleDescriptor tm = new TenantModuleDescriptor();
+    tm.setAction(TenantModuleDescriptor.Action.enable);
+    tm.setId(mdD110.getId());
+    tml.add(tm);
+
+    DepResolution.installSimulate(modsAvailable, modsEnabled, tml, res -> {
+      context.assertTrue(res.succeeded());
+      logger.debug("tml result = " + Json.encodePrettily(tml));
+      context.assertEquals(2, tml.size());
+      context.assertEquals("moduleA-1.1.0", tml.get(0).getId());
+      context.assertEquals("moduleA-1.0.0", tml.get(0).getFrom());
+      context.assertEquals("enable", tml.get(0).getAction().name());
+      context.assertEquals("moduleD-1.1.0", tml.get(1).getId());
+      context.assertEquals(null, tml.get(1).getFrom());
+      context.assertEquals("enable", tml.get(1).getAction().name());
+      async.complete();
+    });
+  }
```

Comment:
There are two relevant cases that should also been tested:
If mdA100, mdA110, mdD100 and mdD110 are available, and mdA100 and mdD100 are enabled.
First case: Enable mdA110 and check that both are upgraded.
Second case: Enable mdD110 and check that both are upgraded.

Reply:
First case: only mdA110 will be upgraded, because mdD100 can use the provided interface of mdA110.

Second case: both will be upgraded, because mdD110 will require mdA110.